=================================================

Milestone 3

Nama  : Jonathan Tuahta

Batch : BSD-02

Objective : 

- Membuat expectation suite dengan kriteria :
1. memastikan unique values
2. memastikan minimal dan maksimal pada kolom
3. memastikan set pada kolom
4. memastikan isi kolom sesuai dengan kriteria
5. kolom numerical
6. kolom categorical
7. memastikan nilai median pada sebuah kolom

=================================================

# A. Input Library

In [1]:
# Install the library

from great_expectations.data_context import FileDataContext

In [2]:
# Create a data context
context = FileDataContext.create(project_root_dir='./')

# B. Connect to A Datasource

In [3]:
# Give a name to a Datasource. This name must be unique between Datasources.
datasource_name = 'csv-Electric-car'
datasource = context.sources.add_pandas(datasource_name)

# Give a name to a data asset
asset_name = 'Electric_car_data'
path_to_data = r"D:\data saintis\Hacktive8_02\phase 2\week2\jonathan\P2M3_Jonathan_Tuahta_clean.csv"
asset = datasource.add_csv_asset(asset_name, filepath_or_buffer=path_to_data)

# Build batch request
batch_request = asset.build_batch_request()

# C. Create an Expectation Suite

In [4]:
# Creat an expectation suite
expectation_suite_name = 'expectation-electric-dataset'
context.add_or_update_expectation_suite(expectation_suite_name)

# Create a validator using above expectation suite
validator = context.get_validator(
    batch_request = batch_request,
    expectation_suite_name = expectation_suite_name
)

# Check the validator
validator.head()

Calculating Metrics:   0%|          | 0/1 [00:00<?, ?it/s]

Unnamed: 0  VIN_(1-10)   country               city state  postal_code  \
0           0  2C4RC1N71H    Kitsap          Bremerton    WA      98311.0   
1           1  2C4RC1N7XL   Stevens           Colville    WA      99114.0   
2           2  KNDC3DLCXN    Yakima             Yakima    WA      98908.0   
3           3  5YJ3E1EA0J    Kitsap  Bainbridge Island    WA      98110.0   
4           4  1N4AZ1CP7J  Thurston           Tumwater    WA      98501.0   

   model_year      make     model                   electric_vehicle_type  \
0        2017  CHRYSLER  PACIFICA  Plug-in Hybrid Electric Vehicle (PHEV)   
1        2020  CHRYSLER  PACIFICA  Plug-in Hybrid Electric Vehicle (PHEV)   
2        2022       KIA       EV6          Battery Electric Vehicle (BEV)   
3        2018     TESLA   MODEL 3          Battery Electric Vehicle (BEV)   
4        2018    NISSAN      LEAF          Battery Electric Vehicle (BEV)   

          clean_alternative_fuel_vehicle_eligibility  electric_range  \
0            Clean Alternative Fuel Vehicle Eligible              33   
1            Clean Alternative Fuel Vehicle Eligible              32   
2  Eligibility unknown as battery range has not b...               0   
3            Clean Alternative Fuel Vehicle Eligible             215   
4            Clean Alternative Fuel Vehicle Eligible             151   

   base_MSRP  legislative_district  DOL_vehicle_ID  \
0          0                  23.0       349437882   
1          0                   7.0       154690532   
2          0                  14.0       219969144   
3          0                  23.0       476786887   
4          0                  35.0       201185253   

                  vehicle_location        electric_utility  2020_census_tract  
0  POINT (-122.6466274 47.6341188)  PUGET SOUND ENERGY INC       5.303509e+10  
1     POINT (-117.90431 48.547075)             AVISTA CORP       5.306595e+10  
2  POINT (-120.6027202 46.5965625)              PACIFICORP       5.307700e+10  
3  POINT (-122.5235781 47.6293323)  PUGET SOUND ENERGY INC       5.303509e+10  
4     POINT (-122.89692 47.043535)  PUGET SOUND ENERGY INC       5.306701e+10

# D.1 - Expectations

## 1. To be unique

In [5]:
# Ekspektasi: Kolom 'customer_id' harus unik
validator.expect_column_values_to_be_unique('DOL_vehicle_ID')

Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

{
  "success": true,
  "result": {
    "element_count": 74986,
    "unexpected_count": 0,
    "unexpected_percent": 0.0,
    "partial_unexpected_list": [],
    "missing_count": 0,
    "missing_percent": 0.0,
    "unexpected_percent_total": 0.0,
    "unexpected_percent_nonmissing": 0.0
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

in column `DOL_vehicle_ID` indicates that all of the values are unique with count of all values are 74986. 

## 2. To be between min_value and max_value

In [6]:
# Ekspektasi: Kolom 'age' harus berada dalam rentang 18 hingga 60 tahun

validator.expect_column_values_to_be_between('electric_range', min_value=0, max_value=337)


Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

{
  "success": true,
  "result": {
    "element_count": 74986,
    "unexpected_count": 0,
    "unexpected_percent": 0.0,
    "partial_unexpected_list": [],
    "missing_count": 0,
    "missing_percent": 0.0,
    "unexpected_percent_total": 0.0,
    "unexpected_percent_nonmissing": 0.0
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

in column `electric_range`, indicates that the values are between 0 to 337.

## 3. To be in set

In [8]:
# Ekspektasi: Kolom 'gender' harus memiliki nilai yang termasuk dalam set ('Male', 'Female', 'Other')
validator.expect_column_values_to_be_in_set('electric_vehicle_type', ['Plug-in Hybrid Electric Vehicle (PHEV)', 'Battery Electric Vehicle (BEV)'])


Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

{
  "success": true,
  "result": {
    "element_count": 74986,
    "unexpected_count": 0,
    "unexpected_percent": 0.0,
    "partial_unexpected_list": [],
    "missing_count": 0,
    "missing_percent": 0.0,
    "unexpected_percent_total": 0.0,
    "unexpected_percent_nonmissing": 0.0
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

from this expectation we know that column `electric_vehicle_type` contains only:

1. plug-in hybrid electric vehicle (PHEV)
2. battery electric vehicle (BEV)

## 4. To be in type list

In [7]:
# Ekspektasi: Kolom 'total_spend' tidak boleh mengandung nilai yang hilang (NULL)
validator.expect_column_values_to_not_be_null('clean_alternative_fuel_vehicle_eligibility')


Calculating Metrics:   0%|          | 0/6 [00:00<?, ?it/s]

{
  "success": true,
  "result": {
    "element_count": 74986,
    "unexpected_count": 0,
    "unexpected_percent": 0.0,
    "partial_unexpected_list": []
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

from this expectation we know that column `clean_alternative_fuel_vehicle_eligibility` have not contain missing values.

## 5. Must be numeric

In [9]:
# Memastikan bahwa semua nilai dalam setiap kolom adalah numerik (integer atau float).
validator.expect_column_values_to_match_regex('legislative_district', r'^\d+\.?\d*$')


Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

{
  "success": true,
  "result": {
    "element_count": 74986,
    "unexpected_count": 0,
    "unexpected_percent": 0.0,
    "partial_unexpected_list": [],
    "missing_count": 0,
    "missing_percent": 0.0,
    "unexpected_percent_total": 0.0,
    "unexpected_percent_nonmissing": 0.0
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

## 6. Must be str

In [10]:
# Ekspektasi : Semua nilai dalam kolom `city` harus terdiri dari type str

validator.expect_column_values_to_be_in_type_list('city', type_list=['str'])


Calculating Metrics:   0%|          | 0/8 [00:00<?, ?it/s]

{
  "success": true,
  "result": {
    "element_count": 74986,
    "unexpected_count": 0,
    "unexpected_percent": 0.0,
    "partial_unexpected_list": [],
    "missing_count": 0,
    "missing_percent": 0.0,
    "unexpected_percent_total": 0.0,
    "unexpected_percent_nonmissing": 0.0
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

## 7. Tobe Betwen Median

In [11]:
# Ekspektasi : Median dari seluruh data pada kolom average_rating adalah 4
validator.expect_column_median_to_be_between('model_year', value = 2021)

Calculating Metrics:   0%|          | 0/4 [00:00<?, ?it/s]

{
  "success": true,
  "result": {
    "observed_value": 2021.0
  },
  "meta": {},
  "exception_info": {
    "raised_exception": false,
    "exception_traceback": null,
    "exception_message": null
  }
}

# E. Save into Expectation Suite


In [12]:
# Save into Expectation Suite

validator.save_expectation_suite(discard_failed_expectations=False)

# F Conclusion
1. semua validasi yang diberikan sesuai dengan keinginan 
2. tujuan validasi ini adalah mempertahankan kualitas dari data agar data sendiri tidak mengalami kekurangan informasi
3. hampir semua kolom digunakan sebagai test pada notebook ini
